[WebAssembly] Lower fmuladd to madd and nmadd

sparker-arm opened the pull request:
Lower v8f16, v4f32 and v2f64 fmuladd calls to fma, when we have relaxed simd.

sparker-arm commented:
The pattern matching confused me yesterday and, after looking with fresh eyes, I'm pretty sure that's because the operands were ordered incorrectly in the original patterns. AFAICT the wasm instructions and ISD nodes both use the same operand order for FMA. @badumbatish Could you please take a look and confirm?

sparker-arm commented:
Performance wise, this looks really good for ML. Running NCNN, via V8 on AArch64, I observe a mean execution time reduction of ~15% over >30 workloads.

badumbatish commented on llvm/test/CodeGen/WebAssembly/simd-relaxed-fma.ll:
```diff
@@ -103,9 +136,8 @@ define <4 x float> @fmuladd_4xf32(<4 x float> %a, <4 x float> %b, <4 x float> %c
 ; RELAXED-LABEL: fmuladd_4xf32:
 ; RELAXED:         .functype fmuladd_4xf32 (v128, v128, v128) -> (v128)
 ; RELAXED-NEXT:  # %bb.0:
-; RELAXED-NEXT:    f32x4.mul $push0=, $0, $1
-; RELAXED-NEXT:    f32x4.add $push1=, $pop0, $2
```
im curious is there a reason why we remove the fmul_contract and fsub_contract?

badumbatish replied:
oh i see, i'm also not quite sure how to reconcile with the relaxed-simd madd's optional round once or twice. if we're going from fma (i assume is just rounding once) to relaxed madd (optionally rounding once or twice), and the platform that runs relaxed madd can only support rounding twice, isn't that some what dangerous?

sparker-arm replied:
Yes, the ISD FMA follows the IEEE convention of rounding only once, so I understand the concern.

The key thing here is that we (backend developers) follow the semantics of the LLVM langref, which allows the fmuladd to either be lowered to a fused op or not, and obey the user.

Whether a fmuladd call is generated in the first place is determined by the language semantics, and frontend options provided by the user.

As noted by @dschuff [here](https://github.com/llvm/llvm-project/issues/55932#issuecomment-3285922007), for C/C++ the operations can be fused in a single statement, without providing additional flags to relax FP constraints.

WebAssembly adds a further restriction though because we only allow the fusing in the presence of relaxed-simd, and this is something that the user has to manually opt-in for. The user needs to understand that the precision of these operations are no longer deterministic when using relaxed-simd.

dschuff replied:
My reading of the LLVM [FMA intrinsic](https://llvm.org/docs/LangRef.html#llvm-fma-intrinsic) is that it only allows double-rounding if the 'afn' flag is set, and it looks like the ISD FMA is the same. That means that the relaxed madd is not actually a valid implementation of that, because it might round once or twice. So in principle, this pattern should only match when the fast math flag is set on the ISD node. (And unfortunately I think that just because the user enabled relaxed-simd doesn't necessarily mean that they are opting in to all FMA operations being allowed to double round. All the feature-enablement flags just control what proposal is allowed in the output, rather than semantics).

So relaxed-simd or no, we have the general problem that if single-rounding is required, there really is no guaranteed valid implementation, short of some kind of slow software emulation.
If the FMAs are generated from e.g. C++ where double or single rounding is allowed, the mul + add or whatever fused operations are generated should already have the appropriate flags and things should just work. (or we could in principle fix them so they do). But something like std::fma requires single rounding, and I don't think we can claim to correctly implement that with relaxed simd without some way for the user to opt-in to the nondeterminism. Which, probably we should go ahead and add somehow, because as you found above, the performance benefit seems pretty substantial.

dschuff replied:
Actually I guess we don't know how much of the performance benefit above comes from fusing mul+add operations that would be valid either way according to C++, and how much of it is from using the wasm instruction for FMAs that are in the source and are required to be fused, right?

sparker-arm replied:
AFAICT, the code base is just relying on the C/C++ semantics. The IR I noticed included `fmuladd` and not `fma`, and there's no special constraints attached the `fmuladd` call.

I will look into an alternative route of lowering the `fmuladd`.

sparker-arm replied:
Okay, I've now introduced a FMULADD node that we can perform pattern match on directly.

badumbatish reviewed (approved):
> The pattern matching confused me yesterday and, after looking with fresh eyes, I'm pretty sure that's because the operands were ordered incorrectly in the original patterns. AFAICT the wasm instructions and ISD nodes both use the same operand order for FMA. @badumbatish Could you please take a look and confirm?
> 
> edit: and changing this fixed a NaN error I was encountering with a workload.

can confirmed it was an ordering mistake. Both wasm fmadd and fma should follow IREE754.

badumbatish commented:
> Performance wise, this looks really good for ML. Running NCNN, via V8 on AArch64, I observe a mean execution time reduction of ~15% over >30 workloads.

is there some sort of generic automated test suite that you're running that you can share with me? for llvm-test-suite, it's a bit hard since cmake can't build the targets (missing libs) so it requires some manual scripting

sparker-arm committed:
[WebAssembly] relaxed madd, nmadd from fmuladd.

Introduce an fmuladd ISD node as an equivalent to the LLVM intrinsic
and lower this to madd and nmadd, when we have relaxed-simd.

sparker-arm committed:
formatting

sparker-arm commented:
Results still look good:

```
Metric                    Execution Time Change(%)
----------------------  --------------------------
Mean (filtered)                             14.854
Median (filtered)                           16.472
Max (filtered)                              25.636
Min (filtered)                               0.015
Geomean (non-filtered)                      13.953 

Benchmark                             Execution Time Change(%)
----------------------------------  --------------------------
ncnn-FastestDet-run_times                               11.111
ncnn-alexnet-run_times                                  16.683
ncnn-blazeface-run_times                                12.803
ncnn-efficientnet_b0-run_times                          11.927
ncnn-efficientnetv2_b0-run_times                        17.4
ncnn-googlenet-run_times                                18.773
ncnn-googlenet_int8-run_times                            0.627
ncnn-mnasnet-run_times                                  17.153
ncnn-mobilenet-run_times                                25.044
ncnn-mobilenet_ssd-run_times                            24.454
ncnn-mobilenet_ssd_int8-run_times                        0.015
ncnn-mobilenet_v2-run_times                             19.737
ncnn-mobilenet_v3-run_times                             14.865
ncnn-mobilenet_yolo-run_times                           25.636
ncnn-mobilenetv2_yolov3-run_times                       21.615
ncnn-nanodet_m-run_times                                18.885
ncnn-proxylessnasnet-run_times                          14.983
ncnn-regnety_400m-run_times                             16.26
ncnn-resnet18-run_times                                 19.614
ncnn-resnet50-run_times                                 23.153
ncnn-shufflenet-run_times                               15.25
ncnn-shufflenet_v2-run_times                            20.588
ncnn-squeezenet-run_times                               19.472
ncnn-squeezenet_int8-run_times                           0.451
ncnn-squeezenet_ssd-run_times                           13.072
ncnn-squeezenet_ssd_int8-run_times                       0.027
ncnn-vgg16-run_times                                    18.597
ncnn-vgg16_int8-run_times                                0.132
ncnn-vision_transformer-run_times                       13.76
ncnn-yolo_fastest_1.1-run_times                         12.267
ncnn-yolo_fastestv2-run_times                           12.088
ncnn-yolov4_tiny-run_times                              18.895
```

badumbatish commented on llvm/test/CodeGen/WebAssembly/simd-relaxed-fma.ll:
```diff
@@ -2,9 +2,278 @@
+; STRICT-LABEL: fadd_fmul_contract_f16:
+; STRICT:         .functype fadd_fmul_contract_f16 (f32, f32, f32) -> (f32)
+; STRICT-NEXT:  # %bb.0:
+; STRICT-NEXT:    call $push0=, __truncsfhf2, $0
```
is there a chance this will not stay in f16 since we're not truncating the final result?

sparker-arm replied:
I can only assume it will not stay as f16 but, as wasm doesn't have scalar fp16 support, returning an f32 appears to be the valid thing to do.

badumbatish reviewed (approved):
LGTM overall i just have a question on the truncate and extending for f16 based operations, we should wait for another reviewer

brendandahl commented on llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:
```diff
@@ -317,6 +317,14 @@ WebAssemblyTargetLowering::WebAssemblyTargetLowering(
     }
 
+    if (Subtarget->hasRelaxedSIMD()) {
+      if (Subtarget->hasFP16()) {
```
Yes, though they don't currently have the `relaxed` prefix in the spec. https://github.com/WebAssembly/half-precision/issues/5

dschuff replied:
OK, great. if you look at the tests below, dow all of those conversion libcalls look like something we should be able to eventually lower to instructions?

not necessarily needed for this PR, but something for the future.

brendandahl replied:
I think so. IIRC, since we don't have a fp16 registers some of the instructions weren't considered legal and were expanded.

brendandahl commented on llvm/lib/Target/WebAssembly/WebAssemblyInstrSIMD.td:
```diff
@@ -1597,16 +1597,20 @@ multiclass SIMDMADD<Vec vec, bits<32> simdopA, bits<32> simdopS, list<Predicate>
 defm "" : SIMDMADD<F32x4, 0x105, 0x106, [HasRelaxedSIMD]>;
 defm "" : SIMDMADD<F64x2, 0x107, 0x108, [HasRelaxedSIMD]>;
-defm "" : SIMDMADD<F16x8, 0x14e, 0x14f, [HasFP16]>;
+defm "" : SIMDMADD<F16x8, 0x14e, 0x14f, [HasRelaxedSIMD, HasFP16]>;
```
The relaxed fp16 instructions are not a separate feature. HasFp16 implies the relaxed instructions are available.

dschuff replied:
oh, interesting. Do you happen to know why? since the spec already has a deterministic profile, this would preclude FP16 from being used in that profile unless we break that dependency (or carve out a deterministic-profile subset of FP16, but that sounds like basically the same thing).

brendandahl replied:
Unfortunately, my chat history is gone for this topic... Now that I'm thinking about it more though, I think Ilya mentioned the instructions are actually deterministic (and also why relaxed was not in the name).

If the hardware supports fp16 then you have an fma instruction. If the hardware doesn't support fp16, you do the math as fp32. Then in both cases you only end up with a single rounding.

sparker-arm replied:
> HasFp16 implies the relaxed instructions are available.

That seems like an odd assumption to make. And, unfortunately, the FP16 proposal doesn't seem to have much information in it... but if it is a fused operation, then we can do lowering differently and use the FMA. I'll make the change.

arsenm commented on llvm/include/llvm/CodeGen/ISDOpcodes.h:
```diff
@@ -514,6 +514,9 @@ enum NodeType {
   /// separately rounded operations.
   FMAD,
 
+  /// FMULADD - Performs a * b + c, with, or without, intermediate rounding.
+  FMULADD,
```
I don't understand why you would need this. It makes sense to have the fmuladd in the IR, but in the backend from the context you know the exact choice and should just go to it

arsenm replied:
Oh...wasm has its own fmuladd. Sigh.

sparker-arm committed:
Use FMA for v8f16

sparker-arm committed:
changed simd-encodings test

sparker-arm committed:
extra test and TODO

dschuff reviewed (approved):
Thanks, this looks good.
On FP16 it's good to get this baseline. I think it probably makes sense to defer further optimization work until we get the proposal to phase 2 and have a better idea of exactly which operations it will include (unless we want to do more benchmarking work to inform that decision of course).

sparker-arm committed:
added FMULADD comment#
